collectors/python.d.plugin/openldap/openldap.chart.py

```diff
@@ -0,0 +1,202 @@
+    ),
+    'unbind_operations': (
+        'cn=Unbind,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'add_operations': (
+        'cn=Add,cn=Operations,cn=Monitor', 'monitorOpInitiated',
+    ),
+    'delete_operations':  (
+        'cn=Delete,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'modify_operations': (
+        'cn=Modify,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'compare_operations': (
+        'cn=Compare,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'search_operations': (
+        'cn=Search,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'write_waiters': (
+        'cn=Write,cn=Waiters,cn=Monitor', 'monitorCounter',
+    ),
+    'read_waiters': (
+        'cn=Read,cn=Waiters,cn=Monitor', 'monitorCounter',
+    ),
+}
+
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.order = ORDER
+        self.definitions = CHARTS
+
+        self.server = configuration.get('server', DEFAULT_SERVER)
+        self.port = configuration.get('port', DEFAULT_PORT)
+        self.username = configuration.get('username')
+        self.password = configuration.get('password')
+        self.timeout = configuration.get('timeout', DEFAULT_TIMEOUT)
+
+        self.alive = False
+        self.conn = None
+
+    def disconnect(self):
+        if self.conn:
+            self.conn.unbind()
+            self.conn = None
+            self.alive = False
+
+    def connect(self):
+        try:
+            self.conn = ldap.initialize('ldap://%s:%s' % (self.server, self.port))
+            self.conn.set_option(ldap.OPT_NETWORK_TIMEOUT, self.timeout)
+            self.conn.simple_bind(self.username, self.password)
+        except ldap.LDAPError as error:
+            self.error(error)
+            return False
+
+        self.alive = True
+        return True
+
+    def reconnect(self):
+        self.disconnect()
+        return self.connect()
+
+    def check(self):
+        if not HAS_LDAP:
+            self.error("'python-ldap' package is needed")
+            return None
+
+        return self.connect() and self.get_data()
+
+    def get_data(self):
+        if not self.alive and not self.reconnect():
+            return None
+
+        data = dict()
+        for key in SEARCH_LIST:
+            bind = SEARCH_LIST[key][0]
+            attr = SEARCH_LIST[key][1]
+            try:
+                num = self.conn.search(bind, ldap.SCOPE_BASE, 'objectClass=*', [attr, ])
+                result_type, result_data = self.conn.result(num, 1)
+            except ldap.LDAPError as error:
+                self.error(error)
+                self.alive = False
+                return None
+
+            try:
+                if result_type == 101:
+                    val = int(result_data[0][1].values()[0][0])
+            except ValueError:
```

Comment:
let's log the error if any, i think it will be good for debugin purposes

use `self.debug`

Reply:
If we're catching a `ValueError` for any reason, we should be logging it, because it means something is misbehaving pretty seriously.  Same goes for pretty much anything except module specific errors and some specific expected cases of `OSError`.

So, yes, please make sure this gets logged if it happens.

Reply:
This indexing worries me. I feel like we need `IndexError` here too. I am sure i have mentioned it already

Reply:
I can't tell for certain without getting into the details of the OpenLDAP python module, but it's probably a good idea to catch it here too, provided of course that things are properly logged.

Reply:
Expanding on that a bit, syntax errors are actually a sub-class of `Exception` in Python, so any typos in the code inside a try/except block that catches `Exception` will mask those.

Other things that are also exceptions that should almost never be caught except explicitly include a handful of standard exit conditions (Ctrl-C and `sys.exit()`, though those are only an issue for a bare `except` clause), a number of cases where the interpreter's internal state is not guaranteed safe (`MemoryError` and `SystemError`), and a wide variety of exceptions that can indicate logic errors in the code.